cudax/include/cuda/experimental/__memory_resource/cuda_async_memory_resource.h

```diff
@@ -0,0 +1,294 @@
+//!
+//! .. warning::
+//!
+//!    ``cuda_async_memory_resource`` does not own the pool and it is the responsibility of the user to ensure that the
+//!    lifetime of the pool exceeds the lifetime of the ``cuda_async_memory_resource``
+//!
+//! @endrst
+class cuda_async_memory_resource
+{
+private:
+  ::cudaMemPool_t __pool_;
+
+  //! @brief Checks whether the passed in alignment is valid
+  //! @param __alignment the alignment to check
+  //! @returns true if \p __alignment is valid
+  _CCCL_NODISCARD static constexpr bool __is_valid_alignment(const size_t __alignment) noexcept
+  {
+    return __alignment <= _CUDA_VMR::default_cuda_malloc_alignment
+        && (_CUDA_VMR::default_cuda_malloc_alignment % __alignment == 0);
+  }
+
+  //! @brief  Returns the default ``cudaMemPool_t`` from the current device
+  //! @throws cuda_error if retrieving the default ``cudaMemPool_t`` fails
+  //! @returns The default memory pool of the current device
+  _CCCL_NODISCARD static ::cudaMemPool_t __get_default_mem_pool(const int __device_id)
+  {
+    _LIBCUDACXX_ASSERT(_CUDA_VMR::__device_supports_pools(__device_id), "cudaMallocAsync not supported");
+
+    ::cudaMemPool_t __pool;
+    _CCCL_TRY_CUDA_API(
+      ::cudaDeviceGetDefaultMemPool, "Failed to call cudaDeviceGetDefaultMemPool", &__pool, __device_id);
+    return __pool;
+  }
+
+public:
+  //! @brief Default constructs the cuda_async_memory_resource using the default \c cudaMemPool_t of the default device
+  //! @throws cuda_error if retrieving the default \c cudaMemPool_t fails
+  cuda_async_memory_resource()
+      : __pool_(__get_default_mem_pool(0))
+  {}
+
+  //! @brief Constructs a cuda_async_memory_resource using the default \c cudaMemPool_t of a given device
+  //! @throws cuda_error if retrieving the default \c cudaMemPool_t fails
+  cuda_async_memory_resource(::cuda::experimental::device_ref __device)
+      : __pool_(__get_default_mem_pool(__device.get()))
+  {}
+
+  cuda_async_memory_resource(int)                   = delete;
+  cuda_async_memory_resource(_CUDA_VSTD::nullptr_t) = delete;
+
+  //! @brief  Constructs the cuda_async_memory_resource from a \c cudaMemPool_t.
+  //! @param __pool The \c cudaMemPool_t used to allocate memory.
+  explicit cuda_async_memory_resource(::cudaMemPool_t __pool) noexcept
+      : __pool_(__pool)
+  {}
+
+  //! @brief  Constructs the cuda_async_memory_resource from a \c cuda_memory_pool by calling native_handle()
+  //! @param __pool The \c cuda_memory_pool used to allocate memory.
+  explicit cuda_async_memory_resource(cuda_memory_pool& __pool) noexcept
+      : __pool_(__pool.native_handle())
+  {}
+
+  //! @brief Allocate device memory of size at least \p __bytes via cudaMallocFromPoolAsync.
+  //! @param __bytes The size in bytes of the allocation.
+  //! @param __alignment The requested alignment of the allocation.
+  //! @throws cuda::std::bad_alloc In case of invalid alignment, or
```

Comment:
For discussion: in RMM we found it helpful to users to create an `rmm::out_of_memory` exception that is thrown when we can determine the allocation failure is due to OOM. In the case of CUDA allocation APIs, they return `cudaErrorMemoryAllocation` in this situation, and RMM special cases that.